File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
n, m = list(map(int, input().split()))
visited = [False] * (n + 1)
arr = []


def dfs(count):
    if count == m:
        print(*arr, end='\n')

    for i in range(1, n + 1):
        if not visited[i]:
            visited[i] = True
            arr.append(i)
            dfs(count + 1)
            arr.pop()
            visited[i] = False


dfs(0)

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {

    public static int[] arr;
    public static boolean[] visited;
    public static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        arr = new int[M];
        visited = new boolean[N];
        dfs(N, M, 0);
        System.out.println(sb);
    }

    public static void dfs(int N, int M, int depth) {
        if (depth == M) {
            for (int num : arr) {
                sb.append(num).append(' ');
            }
            sb.append('\n');
            return;
        }

        for (int i = 0; i < N; i++) {
            if (!visited[i]) {
                visited[i] = true;
                arr[depth] = i + 1;
                dfs(N, M, depth + 1);

                visited[i] = false;
            }
        }
    }

}